File: packages/flags-golang/internal/hasher/hasher.go
// Package hasher implements deterministic MurmurHash3 32-bit bucketing.
//
// This implementation is a faithful port of packages/flags-node/src/hasher.ts.
// For ASCII inputs (the only inputs the SDK ever feeds it: salts, namespaces,
// and userIds) it produces byte-for-byte identical bucket numbers as the
// Node.js, browser, and PHP SDKs. Any divergence here would cause users to be
// allocated to different variations across language boundaries.
package hasher

// BucketSpace mirrors signakit.BucketSpace; duplicated here to keep the
// internal package free of cycles.
const BucketSpace = 10000

// Murmur3_32 returns the 32-bit MurmurHash3 of key with the given seed.
//
// Constants and mixing match Austin Appleby's reference and the TypeScript
// SDK's port (c1=0xcc9e2d51, c2=0x1b873593, fmix with 0x85ebca6b / 0xc2b2ae35).
// Like the TS version it consumes the low byte of each rune (charCodeAt & 0xff)
// — for ASCII strings this is identical to []byte(s).
func Murmur3_32(key string, seed uint32) uint32 {
	const (
		c1 uint32 = 0xcc9e2d51
		c2 uint32 = 0x1b873593
		r1        = 15
		r2        = 13
		m  uint32 = 5
		n  uint32 = 0xe6546b64
	)

	hash := seed
	// We hash the low-byte of each character, matching `key.charCodeAt(i) & 0xff`.
	// For pure ASCII this equals []byte(key); for non-ASCII it matches the TS
	// SDK's (lossy) behavior, ensuring cross-SDK agreement.
	b := make([]byte, len(key))
	for i := 0; i < len(key); i++ {
		b[i] = key[i] & 0xff
	}
	length := len(b)
	blocks := length / 4

	for i := 0; i < blocks; i++ {
		k := uint32(b[i*4]) |
			uint32(b[i*4+1])<<8 |
			uint32(b[i*4+2])<<16 |
			uint32(b[i*4+3])<<24

		k *= c1
		k = (k << r1) | (k >> (32 - r1))
		k *= c2

		hash ^= k
		hash = (hash << r2) | (hash >> (32 - r2))
		hash = hash*m + n
	}

	// Tail
	var k uint32
	tailIndex := blocks * 4
	switch length & 3 {
	case 3:
		k ^= uint32(b[tailIndex+2]) << 16
		fallthrough
	case 2:
		k ^= uint32(b[tailIndex+1]) << 8
		fallthrough
	case 1:
		k ^= uint32(b[tailIndex])
		k *= c1
		k = (k << r1) | (k >> (32 - r1))
		k *= c2
		hash ^= k
	}

	// Finalization
	hash ^= uint32(length)
	hash ^= hash >> 16
	hash *= 0x85ebca6b
	hash ^= hash >> 13
	hash *= 0xc2b2ae35
	hash ^= hash >> 16

	return hash
}

// HashToBucket hashes salt+":"+userId into the [0, BucketSpace) range.
func HashToBucket(salt, userID string) int {
	key := salt + ":" + userID
	return int(Murmur3_32(key, 0) % BucketSpace)
}

// HashForTraffic returns the user's bucket in the traffic namespace.
func HashForTraffic(salt, userID string) int {
	return HashToBucket(salt+":traffic", userID)
}

// HashForVariation returns the user's bucket in the variation namespace.
func HashForVariation(salt, userID string) int {
	return HashToBucket(salt+":variation", userID)
}

// HashForDefault returns the user's bucket in the default-allocation namespace.
func HashForDefault(salt, userID string) int {
	return HashToBucket(salt+":default", userID)
}

File: packages/flags-java/src/main/java/com/signakit/flags/hasher/Hasher.java
package com.signakit.flags.hasher;

import com.signakit.flags.Constants;

/**
 * Deterministic MurmurHash3 (32-bit) bucketing.
 *
 * <p>Faithfully ports {@code packages/flags-node/src/hasher.ts}. The TypeScript
 * version operates on {@code String.charCodeAt} (UTF-16 code units, masked to
 * the low byte) — so we mirror that behavior exactly: each Java {@code char}
 * is treated as a 16-bit unit and {@code & 0xff}-masked when packed into the
 * 4-byte block. This keeps bucket numbers identical across SDKs for the same
 * {@code salt:userId} input.
 */
public final class Hasher {
    private Hasher() {}

    private static final int C1 = 0xcc9e2d51;
    private static final int C2 = 0x1b873593;
    private static final int R1 = 15;
    private static final int R2 = 13;
    private static final int M = 5;
    private static final int N = 0xe6546b64;

    /**
     * MurmurHash3 32-bit. Returns the hash as a {@code long} so the caller
     * can treat it as an unsigned 32-bit value when computing buckets.
     */
    public static long murmur3_32(String key, int seed) {
        int hash = seed;
        int len = key.length();
        int blocks = len / 4;

        for (int i = 0; i < blocks; i++) {
            int k =
                    (key.charAt(i * 4) & 0xff)
                            | ((key.charAt(i * 4 + 1) & 0xff) << 8)
                            | ((key.charAt(i * 4 + 2) & 0xff) << 16)
                            | ((key.charAt(i * 4 + 3) & 0xff) << 24);

            k = k * C1;
            k = Integer.rotateLeft(k, R1);
            k = k * C2;

            hash ^= k;
            hash = Integer.rotateLeft(hash, R2);
            hash = hash * M + N;
        }

        int k = 0;
        int tailIndex = blocks * 4;

        switch (len & 3) {
            case 3:
                k ^= (key.charAt(tailIndex + 2) & 0xff) << 16;
                // fallthrough
            case 2:
                k ^= (key.charAt(tailIndex + 1) & 0xff) << 8;
                // fallthrough
            case 1:
                k ^= key.charAt(tailIndex) & 0xff;
                k = k * C1;
                k = Integer.rotateLeft(k, R1);
                k = k * C2;
                hash ^= k;
                break;
            default:
                break;
        }

        hash ^= len;
        hash ^= hash >>> 16;
        hash = hash * 0x85ebca6b;
        hash ^= hash >>> 13;
        hash = hash * 0xc2b2ae35;
        hash ^= hash >>> 16;

        return hash & 0xFFFFFFFFL;
    }

    public static long murmur3_32(String key) {
        return murmur3_32(key, 0);
    }

    public static int hashToBucket(String salt, String userId) {
        long hash = murmur3_32(salt + ":" + userId);
        return (int) (hash % Constants.BUCKET_SPACE);
    }

    public static int hashForTraffic(String salt, String userId) {
        return hashToBucket(salt + ":traffic", userId);
    }

    public static int hashForVariation(String salt, String userId) {
        return hashToBucket(salt + ":variation", userId);
    }

    public static int hashForDefault(String salt, String userId) {
        return hashToBucket(salt + ":default", userId);
    }
}
